test/simple_malloc.cc

```diff
@@ -3,20 +3,186 @@
+  using time_point = std::chrono::time_point<thread_cpu_clock, duration>;
+
+  static constexpr bool is_steady = true;
+
+  static time_point now() noexcept {
+    timespec tp;
+    clock_gettime(CLOCK_THREAD_CPUTIME_ID, &tp);
+    return time_point(duration(std::chrono::seconds{tp.tv_sec} +
+                               std::chrono::nanoseconds{tp.tv_nsec}));
+  }
+};
+
+struct Stats {
+  uint64_t nb_allocations;
+  uint64_t allocated_bytes;
+  std::chrono::nanoseconds wall_time;
+  std::chrono::nanoseconds cpu_time;
+  pid_t tid;
+};
+
+extern "C" DDPROF_NOINLINE void
+do_lot_of_allocations(uint64_t loop_count, std::chrono::microseconds sleep,
+                      std::chrono::microseconds spin,
+                      std::chrono::milliseconds timeout, Stats *stats) {
+  uint64_t nb_alloc{0};
+  uint64_t alloc_bytes{0};
+
+  auto start_time = std::chrono::steady_clock::now();
+  auto end_time = start_time + timeout;
+  auto start_cpu = thread_cpu_clock::now();
+  for (uint64_t i = 0; i < loop_count; ++i) {
     void *p = malloc(1000);
+    ddprof::DoNotOptimize(p);
+    ++nb_alloc;
+    alloc_bytes += 1000;
     void *p2 = realloc(p, 2000);
+    ddprof::DoNotOptimize(p2);
+    ++nb_alloc;
+    alloc_bytes += 2000;
     free(p2);
     void *p3 = calloc(1, 512);
-    (void)p3;
+    ddprof::DoNotOptimize(p3);
+    ++nb_alloc;
+    alloc_bytes += 512;
     free(p3);
-    usleep(1000);
+    if (sleep.count()) {
+      std::this_thread::sleep_for(sleep);
+    }
+    if (spin.count()) {
+      auto target_time = std::chrono::steady_clock::now() + spin;
+      do {
+        volatile uint64_t sum = 1;
+        for (uint64_t j = 0; j < 100; ++j) {
+          sum = std::sqrt(sum) + std::sqrt(sum);
+        }
+      } while (std::chrono::steady_clock::now() < target_time);
+    }
+
+    if (timeout.count() > 0 && std::chrono::steady_clock::now() >= end_time) {
+      break;
+    }
   }
+  auto end_cpu = thread_cpu_clock::now();
+  end_time = std::chrono::steady_clock::now();
+  *stats = {nb_alloc, alloc_bytes, end_time - start_time, end_cpu - start_cpu,
+            ddprof::gettid()};
+}
+
+void print_header() {
+  printf("TestHeaders:%-8s,%-8s,%-14s,%-14s,%-14s,%-14s\n", "PID", "TID",
+         "alloc_samples", "alloc_bytes", "wall_time", "cpu_time");
+}
+
+void print_stats(pid_t pid, const Stats &stats) {
+  printf("TestStats  :%-8d,%-8d,%-14lu,%-14lu,%-14lu,%-14lu\n", pid, stats.tid,
+         stats.nb_allocations, stats.allocated_bytes, stats.wall_time.count(),
+         stats.cpu_time.count());
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    CLI::App app{"Simple allocation test"};
+
+    unsigned int nb_forks{1};
+    unsigned int nb_threads{1};
+    unsigned int sleep_us{0};
+    unsigned int spin_us{0};
+    int timeout_ms = -1;
+    uint64_t loop_count = std::numeric_limits<uint64_t>::max();
+    std::vector<std::string> exec_args;
+
+    app.add_option("--fork", nb_forks, "Number of processes to create")
```

Comment:
`simple` multithreaded, multi-process that can be arranged to run custom workloads 😄